bundles/org.openhab.binding.bmwconnecteddrive/src/main/java/org/openhab/binding/bmwconnecteddrive/internal/action/ChargeProfileActions.java

```diff
@@ -0,0 +1,406 @@
+/**
+ * Copyright (c) 2010-2021 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.bmwconnecteddrive.internal.action;
+
+import static org.openhab.binding.bmwconnecteddrive.internal.utils.ChargeProfileWrapper.ProfileKey.*;
+
+import java.time.DayOfWeek;
+import java.time.LocalTime;
+import java.util.Optional;
+import java.util.Set;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.bmwconnecteddrive.internal.handler.VehicleHandler;
+import org.openhab.binding.bmwconnecteddrive.internal.utils.ChargeProfileWrapper;
+import org.openhab.binding.bmwconnecteddrive.internal.utils.ChargeProfileWrapper.ProfileKey;
+import org.openhab.core.automation.annotation.ActionInput;
+import org.openhab.core.automation.annotation.ActionOutput;
+import org.openhab.core.automation.annotation.RuleAction;
+import org.openhab.core.thing.binding.ThingActions;
+import org.openhab.core.thing.binding.ThingActionsScope;
+import org.openhab.core.thing.binding.ThingHandler;
+
+/**
+ * The {@link ChargeProfileActions} provides actions for VehicleHandler
+ *
+ * @author Norbert Truchsess - Initial contribution
+ */
+@ThingActionsScope(name = "chargeprofile")
```

Comment:
That implies you could have only a single action per binding. Is this the intention? Or is there a naming-scheme if you want to have multiple actions for the same binding?

Reply:
I'm not aware of a binding that has multiple Action classes for the same Thing. If I see correctly this binding has only one Action class. Is this a hypothetical question or is there maybe some misunderstanding?

Reply:
That is not a hypothetical question - from my point of view it makes sense to have multiple actions per thing but I don't know whether this is somehow forbidden by the OH3 design. Technically it doesn't seem to be. Is there some documentation that lists such conventions? By just looking into the examples this is not obvious.
Of course I can easily change that if multiple actions per thing are not something anybody would ever want in OH.

Reply:
Just for clarification, there can be multiple Actions per Thing, but if I understand correctly, you want to add multiple Action *classes* per Thing?

Reply:
sure - for what other reason does ThingHandler.getServices() return a collection if not to allow for multiple Action-classes per thing?
 public Collection<Class<? extends ThingHandlerService>> getServices();

What would the ThingActionsScope-parameter be good for if not to differentiate in between multiple implementaions of ThingActions? If the scope would be a semantical synonym to the Thing-type then it would be just redundant to the Things id that is also passed to getActions in a script.

Reply:
The method signature should a strong evidence, that the framework supports multiple action classes per Thing! :)

The scope's name will be used to group the actions in the [UI](https://github.com/openhab/openhab-webui/pull/336#issuecomment-692919387). So, if you call it "chargeprofile", it might be a bit lost in the UI.

Reply:
I was not aware it shows in the ui - given this it makes sense to use a common prefix for all thing's actions. Like bmw-charging, bmw-navigation (to send coordinates to the car)

Reply:
ATM all actions are grouped by the binding (which is identical to the scope). Can you elaborate what's your concern about this approach?

Reply:
if a binding has multiple ThingActions-classes and thus multiple scopes those are not grouped beneath each other unless their scopes share a common prefix, don't they?. But the information a common prefix for all ThingAction-classes belonging to the same Thing contains is redundant to the fact the actions belong to the same thing.
And actually this is not just a hypothetical question. I'm planning to add a ThingActions-class to the bmwconnecteddrive-binding for creation and sending of messages containing navigation-destinations to the car. I do not want to prefix those names just for the UI, I'll use scope names without prefix and the UI might then be changed to accommodate this.

Reply:
> if a binding has multiple ThingActions-classes and thus multiple scopes those are not grouped beneath each other unless their scopes share a common prefix, don't they?

I think so, too.

You could file an issue for the UI and core to group the actions by binding. I would check beforehand, if a grouping PR is accepted as we might not discussed the whole story here. I'm not familiar with the insights in core and UI at this place.

Reply:
I'm a bit puzzeld here. I checked rules creation dialog but didn't find any actions from other bindings. I think just the _core actions_ are in. Besides the BMW binding there's also mqtt active and I would expect the _publishMQTTMessage_ here.
![image](https://user-images.githubusercontent.com/20279271/113322275-c8a7c800-9314-11eb-8593-e5612e48c538.png)

Reply:
I agree with @ntruchsess that more actions will come. I've already _sendPointOfInterest_ in mind which  can be sent to the vehicle and has nothing to do with the _ChargeProfile_.
I searched through all present bindings and found only 2 with more than one Action
* [caddx](https://www.openhab.org/addons/bindings/caddx/) - same naming but on different things - doesn't apply in this case
* [max](https://www.openhab.org/addons/bindings/max/) - has 2 different actions _max-cube_ and _max-devices_

I think _chargeprofile_ is too generic because it can apply to several upcoming car bindings in the future. Is it a valid approach to add the binding name as prefix like the _MAX_ binding?
@fwolter Can you give advice here?

Reply:
I think so. But I would use a `.` to separate the binding name from the topic.

Reply:
the scope is valid only for the given thing. So even if multiple bindings use same scopes technically there will be no clash. (Of course any ui that groups 'by scope' should take this into account).
From my point of view prepending the scope with the binding-name is redundant. Of course besides having to type more characters it will do no harm though;-)

Reply:
> the scope is valid only for the given thing.

What's the source of this statement?

Reply:
it is in those lengthy [specs](https://github.com/openhab/openhab-core/blob/b1b58436f2fb3aec8f54185c1437e84780f4264d/bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/defaultscope/ScriptThingActions.java#L59) only few people actually read

Reply:
I cannot really understand the discussion here. 🤔 
Why isn't a single action class enough? As @fwolter pointed out, a single action class can hold any number of actions, so why should there be a need to split it?
It is indeed meant to have the scope be identical to the binding id - this makes it simplest for the user as any exception to this rule can easily lead to mistakes and misunderstandings.
The main reason why it is possible to define the scope at all (and not have the binding id applied by the framework automatically) is that there are bindings that consist out of multiple bundles (such as Bluetooth) - in these cases, it must be possible to let the binding extensions also register action classes.

For the BMW binding, I'd strongly suggest to go with a single action class, unless I am missing something here.

Reply:
@kaikreuzer if you don't understand such discussion you should work on the docs. The main reason for this discussion is that there is no documentation regarding this topic (or if there is then it is well hidden) and the source-code doesn't suggest that it should be such either.
Of course - technically there is no restrictions in number of methods in a single action-class. If there shall only be a single action-class per binding we have to rename the class and some of it's methods as e.g. a 'send' or 'cancel' will be ambigous as soon we add further remoting functionality (like sending points-of-interest)

Reply:
I'm glad that things are clear now.

You could file a PR to improve the docs.

Reply:
Looking at the `telegram` binding gives a good impression what happens if an unlimited number of thing actions are implemented in one `ThingActions` class. If everything that is requested in issues (and what is - in most cases - a good idea), we‘ll easily end up with 1k+ lines in a single class file (already 650 lines now). This will end up in unmanageable code (like the `Connection` class in the `amazonechocontrol` binding).

Reply:
> I'm glad that things are clear now.
> 
> You could file a PR to improve the docs.

I don't see how I can improve the docs. If design-decisions and conventions cannot be derived from the code then it must be written by someone who knows the design-decisions. I can only read the code and I've just learned in openhab that is not a reliable source in terms of extracting wanted behaviour.

Reply:
@weymann please merge https://github.com/weymann/openhab-addons/pull/15 so we comply with @kaikreuzer s philosophy

Reply:
That's a valid discusion! I raised the issue https://github.com/openhab/openhab-addons/issues/10510